lock-api/src/test/java/com/palantir/lock/client/RemoteTimeoutRetryerTest.java

```diff
@@ -0,0 +1,164 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.net.SocketTimeoutException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.function.Function;
+
+import org.immutables.value.Value;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+@SuppressWarnings("unchecked") // Mocks with generic types
+public class RemoteTimeoutRetryerTest {
+    private static final int DEFAULT_VALUE = -1;
+    private static final int QUERY_VALUE = 3;
+    private static final int FORTY_TWO = 42;
+    private static final Exception TIMEOUT_EXCEPTION = new RuntimeException(new SocketTimeoutException("timeout"));
+    private static final Exception RUNTIME_EXCEPTION = new RuntimeException();
+
+    private final Clock clock = mock(Clock.class);
+    private final RemoteTimeoutRetryer retryer = new RemoteTimeoutRetryer(clock);
+    private final Function<IntAndDuration, Integer> query = mock(Function.class);
+
+    @Test
+    public void returnsRequestImmediatelyIfSuccessful() {
+        setupDefaultClock();
+        when(query.apply(any())).thenReturn(FORTY_TWO);
+
+        int result = retryForTenMillisAcceptingOnlyFortyTwo();
+        assertThat(result).isEqualTo(FORTY_TWO);
+
+        verify(query).apply(any(IntAndDuration.class));
+        verifyNoMoreInteractions(query);
+    }
+
+    @Test
+    public void returnsDefaultIfAlwaysFailing() {
+        setupDefaultClock();
+        when(query.apply(any())).thenReturn(5);
+
+        int result = retryForTenMillisAcceptingOnlyFortyTwo();
+        assertThat(result).isEqualTo(DEFAULT_VALUE);
+
+        verify(query, times(3)).apply(any(IntAndDuration.class));
+        verifyNoMoreInteractions(query);
+    }
+
+    @Test
+    public void blockingDurationsDecreaseAsTimePasses() {
+        setupDefaultClock();
+        when(query.apply(any())).thenReturn(5);
+
+        retryForTenMillisAcceptingOnlyFortyTwo();
+
+        ArgumentCaptor<IntAndDuration> captor = ArgumentCaptor.forClass(IntAndDuration.class);
+        verify(query, times(3)).apply(captor.capture());
+        assertThat(captor.getAllValues()).containsExactly(
```

Comment:
this is a bit weird, I think if you have to have this, have something along the lines of:


```
List<Duration> durations = captor.getAllValues().stream().map(IntAndDuration::duration).collect(toList());
assertThat(durations)
  .as("duration is decreasing")
  .isSortedAccordingTo(Comparator.reverseOrder())
```

Reply:
I probably should rename the test. Not entirely happy with that assertion: `10, 9, 8` (or for that matter `10, 10, 10`) would pass when they don't.

Reply:
that's fair